src/main/java/String05.java

```diff
@@ -0,0 +1,13 @@
+public class String05 {
+    public String findCharacterAppearMoreTwice(String str) {
+        StringBuilder result = new StringBuilder();
+        str = str.replaceAll(" ", "");
+        while (str.length() > 1) {
+            if (str.lastIndexOf(str.charAt(0)) > 0) {
+                result.append(str.charAt(0));
+            }
```

Comment:
```suggestion
            if (str.lastIndexOf(str.charAt(0)) > 0 && !result.contains(str.charAt(0)) {
                result.append(str.charAt(0));
            }
```

rồi bỏ lệnh sau

Reply:
Theo em thấy cách của em chạy nhanh hơn anh
StringBuilder không có method contains, phải convert về String
Nếu mà xóa dòng dưới đi thì phải dùng vòng for chạy str.length-1 lần

Reply:
Bản thân lệnh 
```java
str = str.replaceAll(str.charAt(0) + "", "");
```
trong vòng lặp là rất tệ rồi bởi vì trong Java (cũng như tuyệt đại đa số các ngôn ngữ lập trình), String là immutable, có nghĩa là không có cách nào thay đổi giá trị của nó cả. Mọi thao tác tưởng chừng như thay đổi giá trị (như replace dưới đây chẳng hạn) thực chất là đi tạo ra 1 String mới, vừa tốn bộ nhớ, vừa mất công xử lý trên tất cả các ký tự trong chuỗi đó. Việc lạm dụng các hàm xử lý chuỗi trong các tình huống phức tạp là không hề tốt.

Khi optimze một bài toán, mình cần chú ý đến độ phức tạp của thuật toán như O(1), O(n), O(n^2).... Chẳng hạn với n = 1000 và thuật toán đang dùng là O(n^2) thì số thao tác cỡ khoảng 1000^2 = 1 triệu thao tác.

Khi đó, nếu dùng vài mẹo để giảm số lần xử lý trên trên cơ sở n, ví dụ từ 1000 --> 900 thì số thao tác là 900 ^ 2 = 810 000 thao tác. Optimize như thế này khá là vô nghĩa (Cái em đang làm và nghĩ là tốt)

Cái mình cần là làm sao đổi từ O(n^2) về còn O(n ln n) hoặc O(n) thì khi đó, n vẫn là 1000 nhưng số thao tác còn 1000 ít hơn rất nhiều so với 810.000

Quay lại code cũ: 

- Vòng while chạy khoảng n lần
- Thao tác lastIndex chạy n lần 
- Hàm replaceAll chạy n lần
- Hàm append hên xui nhưng thường không nhiều lắm


Comment chỉ là nói lên ý tưởng thôi, thực tế thế nào cần check thêm. Ví dụ không có hàm contains thì có thể dùng hàm [indexOf](https://docs.oracle.com/javase/7/docs/api/java/lang/StringBuilder.html#indexOf(java.lang.String))

Reply:
Ok anh

Reply:
Done